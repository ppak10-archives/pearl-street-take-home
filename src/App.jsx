/**
 * App.jsx
 * Main app component file.
 */

// Node Modules
import {useEffect, useState} from 'react';

// Constants
const MIN_LEVELS = 1;
const MAX_LEVELS = 10;

import SAMPLE_DATA from '../sample_data.json';

export default function App() {
  // Hooks
  const [busNumbersSet, setBusNumbersSet] = useState(new Set([]));
  const [levels, setLevels] = useState(MIN_LEVELS);
  const [busNumber, setBusNumber] = useState('');

  useEffect(() => {
    // Maps through list of buses sample data to create set of bus numbers.
    // Also assumes all bus numbers are intended to be unique.
    setBusNumbersSet(SAMPLE_DATA.buses.map((bus) => bus.number));
  }, []);

  // JSX
  const busNumbersDatalistJSX = [...busNumbersSet].map((busNumber) => (
    <option key={busNumber} value={busNumber} />
  ));

  return (
    <div>
      <h1>Rough Graph</h1>
      <fieldset>
        <legend>Bus Number</legend>
        <input
          list="data"
          onChange={(e) => setBusNumber(e.target.value)}
          placeholder="Select Bus Number"
          type="text"
          value={busNumber}
        />
        <datalist id="data">
          {busNumbersDatalistJSX}
        </datalist>
      </fieldset>
      <fieldset>
        <legend>Levels</legend>
        <input
          min={MIN_LEVELS}
          max={MAX_LEVELS}
          name="levels"
          onChange={(e) => setLevels(e.target.value)}
          type="number"
          value={levels}
        />
      </fieldset>
    </div>
  );
}
