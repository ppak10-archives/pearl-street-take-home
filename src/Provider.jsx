/**
 * Provider.jsx
 * Provider component for app.
 */

// Node Modules
import {arrayOf, element, oneOfType} from 'prop-types';
import {useReducer} from 'react';

// Config
import {MIN_LEVELS} from './config';

// Context
import Context from './context';

// Constants
const INITIAL_STATE = {
  busNumbersSet: new Set([]),
  edgeMap: new Map(),
  levels: MIN_LEVELS,
};

const reducer = (state, action) => {
  switch (action.type) {
    case 'SET_BUS_NUMBERS_SET':
      return {
        ...state,
        busNumbersSet: action.payload.busNumbersSet,
      };
    case 'SET_EDGE_MAP':
      return {
        ...state,
        edgeMap: action.payload.edgeMap,
      };
    case 'SET_LEVELS':
      return {
        ...state,
        levels: action.payload.levels,
      };
    default:
      return state;
  }
};

export default function Provider({children}) {
  // Hooks
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);

  return (
    <Context.Provider value={{state, dispatch}}>
      {children}
    </Context.Provider>
  );
}

Provider.propTypes = {
  children: oneOfType([arrayOf(element), element]).isRequired,
};
