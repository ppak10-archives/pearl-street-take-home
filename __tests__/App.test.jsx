/**
 * App.test.jsx
 * Test file for App component
 */

// Node Modules
import renderer from 'react-test-renderer';

// Components
import App from '../src/App';
import Provider from '../src/Provider';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Provider>
        <App />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
