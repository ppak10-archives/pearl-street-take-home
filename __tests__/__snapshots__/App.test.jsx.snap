// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <h1>
    List view of nodes
  </h1>
  <div>
    <button
      disabled={true}
      onClick={[Function]}
    >
      List
    </button>
    <button
      disabled={false}
      onClick={[Function]}
    >
      Graph
    </button>
  </div>
  <fieldset>
    <legend>
      Bus Number
    </legend>
    <input
      list="data"
      onChange={[Function]}
      placeholder="Select Bus Number"
      type="text"
      value=""
    />
    <datalist
      id="data"
    />
  </fieldset>
  <fieldset>
    <legend>
      Levels
    </legend>
    <input
      max={5}
      min={1}
      onChange={[Function]}
      type="number"
      value={1}
    />
    <input
      max={5}
      min={1}
      onChange={[Function]}
      type="range"
      value={1}
    />
  </fieldset>
  <ul>
    <del />
    <ul />
  </ul>
</div>
`;
